### PR TITLE
Pull request for review by @lauraxt

### PR DESCRIPTION
**Pull request by gregglind**
Laura:

Concerns
- memory usage, leaks
- style / idiom
- reinventing the wheel.
  - function args validators
  - event handnling systems
  - wrapping localstorage
- documentation.

I claim this is a working deployable candidate.

### CODE REVIEW

File: ABOUT.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# About Self-Repair-Server
```
Comment (gregglind) [inline ABOUT.md]: this all needs a review for clarity and content.

File: src/runner.js [hunk 2]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
 let chai = require("chai");
+let type = require("./jetpack/type");
+
+const config = {
+  alwaysRun: false
+};
 
-let config = exports.config = {};
+
+let validate = function (thing, rules) {
```
Comment (gregglind) [inline src/runner.js]: use chai here is overkill.  pick a contracts / validation system somewhere and use it everywhere.

File: src/runner.js [hunk 3]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
-  let errors  = [];
   let rules = [
     function (c) {chai.expect(c.name).to.be.a("string")},
     function (c) {chai.expect(c.description).to.be.a("string")},
-    function (c) {chai.expect(c.recipe).to.be.a("function")},
+    function (c) {chai.expect(c.owner).to.be.a("string")},
+    function (c) {chai.expect(c.version).to.be.a("number")},
+    function (c) {chai.expect(c.run).to.be.a("function")},
     function (c) {chai.expect(c.shouldRun).to.be.a("function")}
   ];
```
Comment (gregglind) [inline src/runner.js]: should we add / require a 'guarantee'  or 'rollback', per @lauraxt.  
implied:
- `bool guarantee(state)` <- the state changed?  this is the claim that the recipe makes.  See #33

File: src/runner.js [hunk 4]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
-
-  return ([errors, errors.length == 0])
+  return validate(config, rules);
   // has keys
   // these are callables?
   //
 }; //
 
 
+/** out guard
```
Comment (gregglind) [inline src/runner.js]: these are *contracts*.  Guard is overloaded in js

File: src/runner.js [hunk 5]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
+      };
+
+      try {
+        recipe.run(state).then(
+          good,
+          bad
+        ); // calls back?  TODO, wants to be promise?
+      } catch (err) {
+        // TODO, this is super bad.
+        reject({name:recipe.name, status: "exception"});
```
Comment (gregglind) [inline src/runner.js]: should this be a reject?  The runner handles it, I think!

File: src/runner.js [hunk 6]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
+    // todo, should we re-get state?  TODO yes.
+    P = P.then(
+      // is this where to re-get state?
+      () => {
+        return statePromiseFn().then(
+          (state) => attemptRun(repair, state, myconfig)
+        );
+      }
+    ).then(
+      record,
```
Comment (gregglind) [inline src/runner.js]: this is where the `guarantee` should fit.  #33

File: src/runner.js [hunk 7]
```diff
@@ -41,74 +55,180 @@ TODO?  Promises or callback as first arg?  Are we node?
...
+      (obj) => {
+        record(obj);  // TODO. badness should get extra reported
```
Comment (gregglind) [inline src/runner.js]: record is a little weak right now.

File: src/main.js [hunk 8]
```diff
@@ -114,7 +140,18 @@ for (let key in runtimeConfig) {
   }
 }
 // actually run
-mainloop(require("./repairs"));
+
+let recipes = require("./repairs")
+mainloop(recipes);
+
+
+window.heartbeat = {
```
Comment (gregglind) [inline src/main.js]: Add comment that this is the only think that should leak to the window. #34

File: src/main.js [hunk 9]
```diff
@@ -80,19 +106,19 @@ let state;
...
 };
 
 
+let guessedLocale = guessLocale();
+if (guessLocale) {
+  personinfo.config.overrides.locale = guessedLocale;
+} // use this locale first.
 
 // process config.  TODO, use a lib for this?
 for (let key in runtimeConfig) {
```
Comment (gregglind) [inline src/main.js]: should we disable this on the build / deployed product?  #36

File: src/common/personinfo.js [hunk 10]
```diff
@@ -13,54 +13,93 @@
...
+  * tour: a tourUi object
+  *
+  * aConfig:  (extends module config)
+  * - overrides:  {}  // note, only common keys will override.
+  *                   // this is not an 'extras'
+  * - timeout (time to wait)
+  *
+  */
+
+let personinfo = function (tour, aConfig) {
```
Comment (gregglind) [inline src/common/personinfo.js]: from @lauraxt:  where does it go?  What goes?  #37

File: src/common/heartbeat-phonehome.js [hunk 11]
```diff
@@ -43,7 +41,7 @@ let config = exports.config  = {
   *        if this is unwanted:  `annotate(extend({},obj))`
   *
   */
-let annotate = exports.annotate = function (obj) {
+let annotate = function (obj) {
```
Comment (gregglind) [inline src/common/heartbeat-phonehome.js]: from @lauraxt: where does it go? What goes? #37

File: src/main.js [hunk 12]
```diff
@@ -80,19 +106,19 @@ let state;
...
-  personinfo.personinfo(
-    null, // tour
-    function (gottenState) {
-      actions.log("about to runAll");
-      console.log(gottenState);
-      state = gottenState;  // leak it
-      runner.runAll(repairsList, state,
-        function () { actions.log("runAll callback"); }
-    );}
+  runner.runAll(repairsList,
```
Comment (gregglind) [inline src/main.js]: should check that repairsList is valid.
- names unique   #38
- all are valid recipes.

**Comment by gregglind** [comment]
I landed this (without any recipes).  So, the WEIGHT IS ON ME.

File: README.md [hunk 13]
```diff
@@ -55,6 +55,118 @@ Build using webpack
...
+- remember that your paths to `common` and such are relative.
+- it's okay to `require` new `node` packages, but these will be reviewed.
+
+
+### Test All The Things
+
+#### Automated
+
+- tests in `/tests/recipes/<yourplace>` to mirror your structure.
+- run tests `npm run test:unit --karmafiles="test/**/*.js"
```
Comment (potch) [inline README.md]: missing a close quote here :)

**Pull request merged**